master/internal/db/postgres.go

```diff
@@ -1441,6 +1452,211 @@ func (db *PgDB) UpdateStep(
+			Metrics: map[string]interface{}{
+				// TODO(brad): Is it OK to drop num_inputs?
+				"validation_metrics": m.Metrics,
+			},
+			TotalBatches: int(m.BatchesTrained),
+			TotalInputs:  int(m.RecordsTrained),
+		}); err != nil {
+			return errors.Wrap(err, "inserting validation metrics")
+		}
+		return nil
+	})
+}
+
+// AddCheckpointMetadata persists metadata for a completed checkpoint to the database.
+func (db *PgDB) AddCheckpointMetadata(
+	ctx context.Context, m *trialv1.CheckpointMetadata,
+) error {
+	return db.withTransaction("add checkpoint metadata", func(tx *sqlx.Tx) error {
+		if _, err := tx.ExecContext(ctx, `
+UPDATE raw_checkpoints SET archived = true
+WHERE trial_id = $1
+  AND trial_run_id < $2
+  AND total_batches >= $3;
+`, m.TrialId, m.TrialRunId, m.TotalBatches); err != nil {
+			return errors.Wrap(err, "archiving checkpoints")
+		}
+
+		startTime, err := derivePriorWorkloadEndTime(ctx, tx, int(m.TrialId))
+		if err != nil {
+			return err
+		}
+
+		if _, err := tx.NamedExecContext(ctx, `
+INSERT INTO raw_checkpoints
+	(trial_id, trial_run_id, state, start_time, end_time, total_batches,
+	 total_inputs, uuid, resources, framework, format, determined_version)
+VALUES
+	(:trial_id, :trial_run_id, :state, :start_time, now(), :total_batches,
+	 :total_inputs, :uuid, :resources, :framework, :format, :determined_version)
+`, model.Checkpoint{
+			TrialID:    int(m.TrialId),
+			TrialRunID: int(m.TrialRunId),
+			State:      model.CompletedState,
+			StartTime:  startTime,
+			//EndTime           *time.Time `db:"end_time" json:"end_time"`
+			TotalBatches:      int(m.TotalBatches),
+			TotalInputs:       int(m.TotalInputs),
+			UUID:              &m.Uuid,
+			Resources:         model.JSONObjFromMapStringInt64(m.Resources),
+			Framework:         m.Framework,
+			Format:            m.Format,
+			DeterminedVersion: m.DeterminedVersion,
+		}); err != nil {
+			return errors.Wrap(err, "inserting checkpoint metadata")
+		}
+		return nil
+	})
+}
+
+// derivePriorWorkloadEndTime approximates the start time of currently reported metrics since
+// resource allocation uses these times.
+func derivePriorWorkloadEndTime(ctx context.Context, tx *sqlx.Tx, trialID int) (time.Time, error) {
+	var endTime time.Time
+	if err := tx.QueryRowxContext(ctx, `
+SELECT greatest(
+	(SELECT max(end_time) FROM raw_steps WHERE trial_id = $1),
+	(SELECT max(end_time) FROM raw_validations WHERE trial_id = $1),
+	(SELECT max(end_time) FROM raw_checkpoints WHERE trial_id = $1),
+	(
+	    SELECT coalesce(r.start_time, t.start_time)
+		FROM trials t
+		LEFT JOIN trial_runs r ON t.id = r.trial_id
+		WHERE t.id = $1
+	))
+`, trialID).Scan(&endTime); err != nil {
+		return time.Time{}, errors.Wrap(err, "getting prior training metrics")
+	}
+	return endTime, nil
+}
+
+// UpdateStepByPriorBatches updates an existing step. Fields that are nil or zero are not
+// updated.  end_time is set if the step moves to a terminal state.
+func (db *PgDB) UpdateStepByPriorBatches(
```

Comment:
This needs to be removed since we want to be consistent in using `total_batches` as the indexes.